authentication/AuthenticationExampleSwift/MainViewController.swift

```diff
@@ -786,5 +800,110 @@ class MainViewController: UITableViewController {
     }
     tableView.reloadData()
   }
+
+  // MARK: - Sign in with Apple
+
+  // Unhashed nonce.
+  fileprivate var currentNonce: String?
+
+  @available(iOS 13, *)
+  func startSignInWithAppleFlow() {
+    let nonce = randomNonceString()
+    currentNonce = nonce
+    let appleIDProvider = ASAuthorizationAppleIDProvider()
+    let request = appleIDProvider.createRequest()
+    request.requestedScopes = [.fullName, .email]
+    request.nonce = sha256(nonce)
+
+    let authorizationController = ASAuthorizationController(authorizationRequests: [request])
+    authorizationController.delegate = self
+    authorizationController.presentationContextProvider = self
+    authorizationController.performRequests()
+  }
+
+  // Adapted from https://auth0.com/docs/api-auth/tutorials/nonce#generate-a-cryptographically-random-nonce
+  private func randomNonceString(length: Int = 32) -> String {
+    precondition(length > 0)
+    let charset: Array<Character> =
+        Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
+    var result = ""
+    var remainingLength = length
+
+    while remainingLength > 0 {
+      let randoms: [UInt8] = (0 ..< 16).map { _ in
+        var random: UInt8 = 0
+        let errorCode = SecRandomCopyBytes(kSecRandomDefault, 1, &random)
```

Comment:
Can we just generate L bytes and convert to a string? Does the character set matter if we're going to hash and output a hex string anyway?

Reply:
afaik there's no string encoding where all possible values are valid characters. Using a character set is an easy way to avoid this.